app/models/work_packages/scopes/relatable.rb

```diff
@@ -40,33 +40,114 @@ module Relatable
+      # the predecessor's target would not be valid.
+      # It is also important to note, that existing ancestors are of no importance since this part of the structure will change.
+      # Creating a parent relationship is destructive since there can only ever be one.
+      #
+      # The result is a blocklist which will include work packages that can not be related to. The list is not complete
+      # as it will not include the work packages related by different relation types so those are added additionally. For the
+      # PARENT relationship, work packages directly related to any of the descendants are added as well. Since work packages
+      # related to the queried work packages only by a path containing both FOLLOWS and PRECEDES relationships are valid targets,
+      # these are filtered out as well. It would be possible to optimize the CTE here to not have them included
+      # in the first place.
+      #
+      # The CTE has the following columns:
+      # * id - the id of the work packages currently related. This is the result of the CTE.
+      # * from_(hierarchy/from_id/to_id) - booleans to prevent that the CTE returns back the path calculated in the previous
+      #                                    iteration.
+      # * origin - boolean to indicate whether the work package is the queried for work package or its ancestor/descendant
+      #            (only descendant for PARENT). Such a work package is never a valid target.
+      #            Additionally, it is used to limit proceeding in both directions for FOLLOWS/PRECEDES.
+      # * includes_(from_relation/to_relation) - booleans about the direction (from_id -> to_id or to_id -> from_id) of the path
+      #                                          (the relations followed).
+      #                                          This is relevant for a queried for PARENT relation. In that case, relations need
+      #                                          to be followed from the queried for work package (and its descendants) in both
+      #                                          directions. But only the direction taken from that origin needs to be followed
+      #                                          henceforth.
+      # * includes_hierarchy - boolean indicating that the last relation taken was a hierarchy relation. For a queried for
+      #                        PARENT relation, whenever that is the case, the work package is a valid relation target although
+      #                        it appears in the CTE.
       def relatable(work_package, relation_type)
         return all if work_package.new_record?
 
         scope = not_having_directed_relation(work_package, relation_type)
-                  .not_having_direct_relation(work_package)
-                  .where.not(id: work_package)
+                  .not_having_direct_relation(work_package, relation_type)
+                  .where.not(id: work_package.id)
```

Comment:
I would almost define a `not_having` method and use it instead of the `where.not`

```
def not_having(work_package_id)
  where.not(id: work_package_id)
end
```

This way all calls have the same "shape" and the same abstraction level.